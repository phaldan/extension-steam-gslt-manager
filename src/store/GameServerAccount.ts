import { action, observable } from 'mobx';

export default class GameServerAccount {
  public steamid: string;
  public appid: number;

  @observable
  public expired: boolean = false;

  @observable
  public lastLogon?: Date;

  @observable
  public memo: string;

  @observable
  public token: string;

  @observable
  public locked: boolean = false;

  constructor({ steamid, appid, token, memo, lastLogon, expired }) {
    this.steamid = steamid;
    this.appid = appid;
    this.token = token;
    this.memo = memo;
    this.lastLogon = lastLogon ? new Date(lastLogon) : undefined;
    this.expired = expired;
  }

  @action
  public toggleLock(locked: boolean) {
    this.locked = locked;
  }

  @action
  public updateFromJson({ appid, token, memo, lastLogon, expired }) {
    this.appid = appid;
    this.token = token;
    this.memo = memo;
    this.lastLogon = lastLogon;
    this.expired = expired;
  }
}
