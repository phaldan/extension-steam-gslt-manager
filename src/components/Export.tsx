import copy from 'copy-to-clipboard';
import { observer } from 'mobx-react';
import * as React from 'react';
import { Checkbox, FormControl, FormGroup } from 'react-bootstrap';
import GameServerAccount from '../store/GameServerAccount';
import ColumnSelectionState from '../uiState/ColumnSelectionState';
import ActionModal from './ActionModal';

interface ExportProps {
  accounts: GameServerAccount[];
}

@observer
export default class Export extends React.Component<ExportProps> {
  private checked: ColumnSelectionState;

  constructor(props) {
    super(props);
    this.checked = new ColumnSelectionState();
  }

  public render() {
    return (
      <ActionModal
        title="Export"
        buttonIcon="export"
        buttonStyle="primary"
        buttonText="Export"
        size="large"
        action="Copy to clipboard"
        onAction={this.copy}
      >
        <FormGroup>
          {this.createCheckBox(this.checked.steamId, 'js-steamid', 'SteamID', this.setSteamId)}
          {this.createCheckBox(this.checked.token, 'js-token', 'Token', this.setToken)}
          {this.createCheckBox(this.checked.appId, 'js-appid', 'AppID', this.setAppId)}
          {this.createCheckBox(this.checked.lastLogon, 'js-logon', 'Last logon', this.setLogon)}
          {this.createCheckBox(this.checked.memo, 'js-memo', 'Memo', this.setMemo)}
        </FormGroup>

        <FormGroup>
          <FormControl
            componentClass="textarea"
            rows="10"
            readOnly={true}
            value={this.content}
            className="js-output"
          />
        </FormGroup>
      </ActionModal>
    );
  }

  private copy = () => {
    copy(this.content);
  }

  private get content() {
    const { accounts } = this.props;
    return accounts.map((account) => {
      return this.buildColumns(account).join(',');
    }).join('\n');
  }

  private buildColumns(account: GameServerAccount): string[] {
    const columns: string[] = [];
    if (this.checked.steamId) {
        columns.push(account.steamid);
    }
    if (this.checked.token) {
        columns.push(account.token);
    }
    if (this.checked.appId) {
        columns.push(account.appid.toString());
    }
    if (this.checked.lastLogon) {
        columns.push((account.lastLogon) ? account.lastLogon.toISOString() : '');
    }
    if (this.checked.memo) {
        columns.push(account.memo);
    }
    return columns;
  }

  private setSteamId = (event) => {
    this.checked.toggleSteamId(event.target.checked);
  }

  private setToken = (event) => {
    this.checked.toggleToken(event.target.checked);
  }

  private setAppId = (event) => {
    this.checked.toggleAppId(event.target.checked);
  }

  private setLogon = (event) => {
    this.checked.toggleLastLogon(event.target.checked);
  }

  private setMemo = (event) => {
    this.checked.toggleMemo(event.target.checked);
  }

  private createCheckBox(checked: boolean, className: string, text: string, callback) {
    return (
      <Checkbox
        inline={true}
        checked={checked}
        onChange={callback}
        className={className}
      >
       {text}
      </Checkbox>
    );
  }
}
