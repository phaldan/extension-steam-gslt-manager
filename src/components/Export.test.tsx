import { mount, shallow } from 'enzyme';
import * as React from 'react';
import GameServerAccount from '../store/GameServerAccount';
import Export from './Export';

jest.mock('copy-to-clipboard');

describe('<Export>', () => {
  const copyToClipboard = require('copy-to-clipboard');

  beforeEach(() => {
    copyToClipboard.mockReset();
  });

  test('empty account list', () => {
    const target = shallow(<Export accounts={[]} />);
    expect(target).toMatchSnapshot();
  });

  test('empty lastlogon', () => {
    const account = new GameServerAccount({
      appid: 740,
      expired: false,
      lastLogon: undefined,
      memo: '',
      steamid: '212V16ECZ4HE',
      token: '7FJS3VY2273L',
    });
    const target = shallow(<Export accounts={[account]} />);
    target.find('.js-logon').simulate('change', { target: { checked: true } });
    expect(target.find('.js-output').props().value).toEqual('7FJS3VY2273L,');
    expect(target).toMatchSnapshot();
  });

  describe('with single account', () => {
    let target;

    beforeEach(() => {
      const account = new GameServerAccount({
        appid: 740,
        expired: false,
        lastLogon: '1995-12-17T03:24:00',
        memo: 'CSGO',
        steamid: '212V16ECZ4HE',
        token: '7FJS3VY2273L',
      });
      target = shallow(<Export accounts={[account]} />);
    });

    test('initial state', () => {
      expect(target.find('.js-output').props().value).toEqual('7FJS3VY2273L');
      expect(copyToClipboard).not.toBeCalled();
      expect(target).toMatchSnapshot();
    });

    test('check SteamID', () => {
      target.find('.js-steamid').simulate('change', { target: { checked: true } });
      expect(target.find('.js-output').props().value).toEqual('212V16ECZ4HE,7FJS3VY2273L');
      expect(copyToClipboard).not.toBeCalled();
      expect(target).toMatchSnapshot();
    });

    test('uncheck token', () => {
      target.find('.js-token').simulate('change', { target: { checked: false } });
      expect(target.find('.js-output').props().value).toEqual('');
      expect(copyToClipboard).not.toBeCalled();
      expect(target).toMatchSnapshot();
    });

    test('check Appid', () => {
      target.find('.js-appid').simulate('change', { target: { checked: true } });
      expect(target.find('.js-output').props().value).toEqual('7FJS3VY2273L,740');
      expect(copyToClipboard).not.toBeCalled();
      expect(target).toMatchSnapshot();
    });

    test('check last logon', () => {
      target.find('.js-logon').simulate('change', { target: { checked: true } });
      expect(target.find('.js-output').props().value).toEqual('7FJS3VY2273L,1995-12-17T03:24:00.000Z');
      expect(copyToClipboard).not.toBeCalled();
      expect(target).toMatchSnapshot();
    });

    test('check memo', () => {
      target.find('.js-memo').simulate('change', { target: { checked: true } });
      expect(target.find('.js-output').props().value).toEqual('7FJS3VY2273L,CSGO');
      expect(copyToClipboard).not.toBeCalled();
      expect(target).toMatchSnapshot();
    });

    test('copy tp clipboard', () => {
      target.find('ActionModal').simulate('action');
      expect(copyToClipboard).toBeCalled();
      expect(copyToClipboard).toHaveBeenLastCalledWith('7FJS3VY2273L');
      expect(target).toMatchSnapshot();
    });
  });
});
