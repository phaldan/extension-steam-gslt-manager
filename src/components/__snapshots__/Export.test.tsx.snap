// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Export> empty account list 1`] = `
<ActionModal
  action="Copy to clipboard"
  buttonIcon="export"
  buttonStyle="primary"
  buttonText="Export"
  close="Abort"
  disabled={false}
  onAction={[Function]}
  size="large"
  title="Export"
>
  <FormGroup
    bsClass="form-group"
  >
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-steamid"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      SteamID
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={true}
      className="js-token"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Token
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-appid"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      AppID
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-logon"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Last logon
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-memo"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Memo
    </Checkbox>
  </FormGroup>
  <FormGroup
    bsClass="form-group"
  >
    <FormControl
      bsClass="form-control"
      className="js-output"
      componentClass="textarea"
      readOnly={true}
      rows="10"
      value=""
    />
  </FormGroup>
</ActionModal>
`;

exports[`<Export> empty lastlogon 1`] = `
<ActionModal
  action="Copy to clipboard"
  buttonIcon="export"
  buttonStyle="primary"
  buttonText="Export"
  close="Abort"
  disabled={false}
  onAction={[Function]}
  size="large"
  title="Export"
>
  <FormGroup
    bsClass="form-group"
  >
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-steamid"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      SteamID
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={true}
      className="js-token"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Token
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-appid"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      AppID
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={true}
      className="js-logon"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Last logon
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-memo"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Memo
    </Checkbox>
  </FormGroup>
  <FormGroup
    bsClass="form-group"
  >
    <FormControl
      bsClass="form-control"
      className="js-output"
      componentClass="textarea"
      readOnly={true}
      rows="10"
      value="7FJS3VY2273L,"
    />
  </FormGroup>
</ActionModal>
`;

exports[`<Export> with single account check Appid 1`] = `
<ActionModal
  action="Copy to clipboard"
  buttonIcon="export"
  buttonStyle="primary"
  buttonText="Export"
  close="Abort"
  disabled={false}
  onAction={[Function]}
  size="large"
  title="Export"
>
  <FormGroup
    bsClass="form-group"
  >
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-steamid"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      SteamID
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={true}
      className="js-token"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Token
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={true}
      className="js-appid"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      AppID
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-logon"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Last logon
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-memo"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Memo
    </Checkbox>
  </FormGroup>
  <FormGroup
    bsClass="form-group"
  >
    <FormControl
      bsClass="form-control"
      className="js-output"
      componentClass="textarea"
      readOnly={true}
      rows="10"
      value="7FJS3VY2273L,740"
    />
  </FormGroup>
</ActionModal>
`;

exports[`<Export> with single account check SteamID 1`] = `
<ActionModal
  action="Copy to clipboard"
  buttonIcon="export"
  buttonStyle="primary"
  buttonText="Export"
  close="Abort"
  disabled={false}
  onAction={[Function]}
  size="large"
  title="Export"
>
  <FormGroup
    bsClass="form-group"
  >
    <Checkbox
      bsClass="checkbox"
      checked={true}
      className="js-steamid"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      SteamID
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={true}
      className="js-token"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Token
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-appid"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      AppID
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-logon"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Last logon
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-memo"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Memo
    </Checkbox>
  </FormGroup>
  <FormGroup
    bsClass="form-group"
  >
    <FormControl
      bsClass="form-control"
      className="js-output"
      componentClass="textarea"
      readOnly={true}
      rows="10"
      value="212V16ECZ4HE,7FJS3VY2273L"
    />
  </FormGroup>
</ActionModal>
`;

exports[`<Export> with single account check last logon 1`] = `
<ActionModal
  action="Copy to clipboard"
  buttonIcon="export"
  buttonStyle="primary"
  buttonText="Export"
  close="Abort"
  disabled={false}
  onAction={[Function]}
  size="large"
  title="Export"
>
  <FormGroup
    bsClass="form-group"
  >
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-steamid"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      SteamID
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={true}
      className="js-token"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Token
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-appid"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      AppID
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={true}
      className="js-logon"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Last logon
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-memo"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Memo
    </Checkbox>
  </FormGroup>
  <FormGroup
    bsClass="form-group"
  >
    <FormControl
      bsClass="form-control"
      className="js-output"
      componentClass="textarea"
      readOnly={true}
      rows="10"
      value="7FJS3VY2273L,1995-12-17T03:24:00.000Z"
    />
  </FormGroup>
</ActionModal>
`;

exports[`<Export> with single account check memo 1`] = `
<ActionModal
  action="Copy to clipboard"
  buttonIcon="export"
  buttonStyle="primary"
  buttonText="Export"
  close="Abort"
  disabled={false}
  onAction={[Function]}
  size="large"
  title="Export"
>
  <FormGroup
    bsClass="form-group"
  >
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-steamid"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      SteamID
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={true}
      className="js-token"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Token
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-appid"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      AppID
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-logon"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Last logon
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={true}
      className="js-memo"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Memo
    </Checkbox>
  </FormGroup>
  <FormGroup
    bsClass="form-group"
  >
    <FormControl
      bsClass="form-control"
      className="js-output"
      componentClass="textarea"
      readOnly={true}
      rows="10"
      value="7FJS3VY2273L,CSGO"
    />
  </FormGroup>
</ActionModal>
`;

exports[`<Export> with single account copy tp clipboard 1`] = `
<ActionModal
  action="Copy to clipboard"
  buttonIcon="export"
  buttonStyle="primary"
  buttonText="Export"
  close="Abort"
  disabled={false}
  onAction={[Function]}
  size="large"
  title="Export"
>
  <FormGroup
    bsClass="form-group"
  >
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-steamid"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      SteamID
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={true}
      className="js-token"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Token
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-appid"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      AppID
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-logon"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Last logon
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-memo"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Memo
    </Checkbox>
  </FormGroup>
  <FormGroup
    bsClass="form-group"
  >
    <FormControl
      bsClass="form-control"
      className="js-output"
      componentClass="textarea"
      readOnly={true}
      rows="10"
      value="7FJS3VY2273L"
    />
  </FormGroup>
</ActionModal>
`;

exports[`<Export> with single account initial state 1`] = `
<ActionModal
  action="Copy to clipboard"
  buttonIcon="export"
  buttonStyle="primary"
  buttonText="Export"
  close="Abort"
  disabled={false}
  onAction={[Function]}
  size="large"
  title="Export"
>
  <FormGroup
    bsClass="form-group"
  >
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-steamid"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      SteamID
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={true}
      className="js-token"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Token
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-appid"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      AppID
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-logon"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Last logon
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-memo"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Memo
    </Checkbox>
  </FormGroup>
  <FormGroup
    bsClass="form-group"
  >
    <FormControl
      bsClass="form-control"
      className="js-output"
      componentClass="textarea"
      readOnly={true}
      rows="10"
      value="7FJS3VY2273L"
    />
  </FormGroup>
</ActionModal>
`;

exports[`<Export> with single account uncheck token 1`] = `
<ActionModal
  action="Copy to clipboard"
  buttonIcon="export"
  buttonStyle="primary"
  buttonText="Export"
  close="Abort"
  disabled={false}
  onAction={[Function]}
  size="large"
  title="Export"
>
  <FormGroup
    bsClass="form-group"
  >
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-steamid"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      SteamID
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-token"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Token
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-appid"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      AppID
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-logon"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Last logon
    </Checkbox>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      className="js-memo"
      disabled={false}
      inline={true}
      onChange={[Function]}
      title=""
    >
      Memo
    </Checkbox>
  </FormGroup>
  <FormGroup
    bsClass="form-group"
  >
    <FormControl
      bsClass="form-control"
      className="js-output"
      componentClass="textarea"
      readOnly={true}
      rows="10"
      value=""
    />
  </FormGroup>
</ActionModal>
`;
